src/nupic/ntypes/Sdr.hpp

```diff
@@ -0,0 +1,273 @@
+     * @returns The total number of boolean values in the SDR.
+     */
+    const UInt getSize() const
+        {return size;};
+
+    /**
+     * Remove the value from this SDR.  Attempting to get the value of an empty
+     * SDR will raise an exception.
+     */
+    void clear();
+
+    /**
+     * Set all of the values in the SDR to false.  This method overwrites the
+     * SDRs current value.
+     */
+    void zero();
+
+    /**
+     * Assigns a new value to the SDR, overwritting the current value.
+     *
+     * @param value A dense array in a vector<char> to assign to the SDR.
+     */
+    void setDense( const vector<Byte> value );
+
+    /**
+     * Assigns a new value to the SDR, overwritting the current value.
+     *
+     * @param value A dense C-style array of UInt's to assign to the SDR.  The
+     * given value must have as many elements as this SDR, as given by 
+     * SDR.getSize().
+     */
+    void setDense( const UInt *value );
+
+    /**
+     * Assigns a new value to the SDR, overwritting the current value.
+     *
+     * @param value A dense byte Array to assign to the SDR.
+     */
+    void setDense( const ArrayBase *value );
+
+    /**
+     * Assigns a vector of sparse indices to the SDR.  These indicies are into
+     * the flattened SDR space.  This overwrites the SDR's current value.
+     *
+     * @param value A vector of indices to assign to the SDR.
+     */
+    void setFlatIndex( const vector<UInt> value );
+
+    /**
+     * Assigns an array of sparse indices to the SDR.  These indicies are into
+     * the flattened SDR space.  This overwrites the SDR's current value.
+     *
+     * @param value A C-style array of indices to assign to the SDR.
+     * @param num_values The number of elements in the 'value' array.
+     */
+    void setFlatIndex( const UInt *value, const UInt num_values );
+
+    /**
+     * Assign a list of indices to the SDR, overwritting the SDRs current value.
+     * These indices are into the SDR space with dimensions.  The outter list
+     * is indexed using an index into the dimensions list, see
+     * SDR.getDimensions().  The inner lists are indexed in parallel, they
+     * contain the coordinates of the true values in the SDR.
+     *
+     * @param value A list of lists containing the coordinates of the true
+     * values to assign to the SDR.
+     */
+    void setIndex( const vector<vector<UInt>> value );
+
+    /**
+     * Assigns a deep copy of the given SDR to this SDR.  This overwrites the
+     * current value of this SDR.
+     *
+     * @param value An SDR to copy the value of.
+     */
+    void assign( const SparseDistributedRepresentation value );
+
+    /**
+     * Gets the current value of the SDR.  The result of this method call is
+     * saved inside of this SDR until the SDRs value changes.  If the value
+     * is unset this raises an exception.
+     *
+     * @returns A list of all values in the SDR.
+     */
+    const vector<Byte> getDense();
```

Comment:
Sorry, I forgot how C++ vectors & pointers work. I've updated the header file to return a pointer to its internal storage vector, instead of a copy.  If you want a raw pointer to the internal storage then you can use:
`const Byte *raw_ptr =  sdr.getDense()->data();`

It's a constant pointer because if you write to it then this class needs to know so that it can update the other data formats.  I could add a method for returning non-constant pointers to the internal data, but you'd need to follow up after any writes by notifying the SDR class that it's internal data has changed.  Something like:
```
SDR x({1000});
x.zero();
vector<Byte> *ptr = x.getDenseMutable();
ptr->at(99) = 1;
x.getFlatIndex()  ->  {}   // At this point the SDR formats are out of sync...
// notify SDR that dense has change so the other formats need to be recalculated.
x.setDenseInplace();
x.getFlatIndex()  ->  {99}  // SDR formats are in sync again.
```

I don't think that this solves the underlying problem of std::vector not interfacing with *strange* data sources.  std::vector doesn't accept pointers, instead it forces a copy into its own internally managed storage space.  I honestly don't think that the overhead of a few copies is going to be an issue ...

Reply:
> `Oh, did you try that trick I suggested about overloading the vector's allocator to deal with strange data sources?`

No, I just rewrote the class to use a `Byte*` for dense storage.  I then added methods to swap-out the SDR's data pointers.  I also added methods to modify the data in-place.  

> `Maintaining multiple internal formats will also incur overhead copying/conversion.`

Yes but by caching the result of every conversion this class should keep that overhead to a minimum, at least in theory.

I think that the performance issues are because the code uses dense SDRs when it should be using sparse SDRs.  Take for example the following code snippet from the SP method updateDutyCyclesHelper_ which performs an exponential rolling average:
```
// newValues is the SDR, contains 1's and 0's
for (UInt i = 0; i < dutyCycles.size(); i++) {
    dutyCycles[i] = (dutyCycles[i] * (period - 1) + newValues[i]) / period;
}
```
This does a lot of floating-point multiplication by 0 and 1, and adding 0's. which can all be optimized away if you have the sparse indices.  Furthermore this method is called 2 times each compute cycle.  This could be better written as:
```
for (UInt i = 0; i < dutyCycles.size(); i++) {
    dutyCycles[i] *= (period - 1)  / period;
}
for(auto i : newValuesSparse) {
    dutyCycles[i] += 1. / period;
}
```

Reply:
Yes, there is a lot of that type of optimization that needs to be done as well.